### PR TITLE
Issue 1652

### LINKED ISSUE #1652: RDKit interpretation of atom stereo SMILES is different from 4 other toolkits
**Description:**

RDKit is interpreting atom stereo SMILES differently than other tools when a substituent precedes a ring closure digit like so.  

```python
>>> rdkit.__version__
'2017.09.1'
>>> Chem.CanonSmiles('F[C@@]1(C)CCO1')
'C[C@]1(F)CCO1'
>>> Chem.CanonSmiles('[C@@](F)1(C)CCO1')
'C[C@@]1(F)CCO1'
```

From my experimenting the problem only seems to occur when all of the following conditions are true:

- The first atom has atom stereo specified
- That atom is in a single ring (not a bridgehead)
- A substituent of that atom stereo comes before the closure digit for the ring

Full discussion is here: https://www.mail-archive.com/rdkit-discuss@lists.sourceforge.net/msg07447.html

The mailing list discussion includes a test suite covering multiple toolkits. Here is the relevant piece for RDKit: 

```python
from __future__ import print_function
import rdkit
from rdkit import Chem

print("RDKit =", rdkit.__version__)

SMILES_THAT_ARE_THE_SAME = [
    # examples from OpenSMILES spec, every toolkits agrees on these at least
    ('N[C@H](O)C',       '[C@@H](N)(O)C'),
    ('N[C@@](Br)(C)O',   'N[C@](Br)(O)C'),
    ('Br[C@](N)(C)O',    'O[C@](Br)(C)N'),

    # examples with attachment points every toolkit agrees on
    ('[*][C@@H](C)N',    '[C@H]([*])(C)N'),
    ('[*][C@@](F)(C)N',  '[C@@]([*])(F)(C)N'),

    # examples from Dalke 2017 UGM talk
    ('[*][C@](N)(O)S', '[C@]([*])(N)(O)S'),
    ('[*][C@H](O)S',   '[C@@H]([*])(O)S'),
    ('[*:1][C@]1([*:2])CC1(Cl)Cl', '[C@]([*:1])1([*:2])CC1(Cl)Cl'), # RDKit thinks these are different, and ChemAxon removes this stereochemistry entirely

    # example from Dalke report here: https://www.mail-archive.com/rdkit-discuss@lists.sourceforge.net/msg05296.html
    ('CN[S@](c1ccccc1)=O', 'CN[S@]2=O.c12ccccc1'),

    # non-ring cases with no hydrogen property
    ('Cl[C@](F)(Br)(O)', '[C@](Cl)(F)(Br)(O)'),

    # OpenBabel and Indigo have issues with sulfur chirality
    ('Cl[S@](C)=O',      '[S@](Cl)(C)=O'),    # ChemAxon flips this chirality
    ('Cl[S@](C)=CCCC',   '[S@](Cl)(C)=CCCC'), # ChemAxon removes this chirality entirely

    # ring cases, RDKit thinks all of these are different isomers
    ('Cl[C@](F)1CC[C@H](F)CC1',    '[C@](Cl)(F)1CC[C@H](F)CC1'),
    ('Cl[C@]1(c2ccccc2)NCCCS1',    '[C@](Cl)1(c2ccccc2)NCCCS1'),
    ('Cl3.[C@]31(c2ccccc2)NCCCS1', '[C@](Cl)1(c2ccccc2)NCCCS1'),
    ('Cl[C@](F)1C2C(C1)CNC2',      '[C@](Cl)(F)1C2C(C1)CNC2'),
    ('[*][C@@H]1CO1',              '[C@H]([*])1CO1'),     # with hydrogen property
    ('[*][C@@]1(C)CCO1',           '[C@@]([*])1(C)CCO1'), # without hydrogen property
    ('F[C@@]1(C)CCO1',             '[C@@](F)1(C)CCO1'), # with a real atom to be sure

    # bridge-head ring case, RDKit agrees with all the other toolkits on this case
    ('[*][C@@]12CNC[C@H]1C2',   '[C@@]([*])12CNC[C@H]1C2'),

    # ring cases from https://github.com/rdkit/rdkit/commit/a2fe13f85a89cd66903ca63edeec0f99b61e8185#diff-6c7d3f98ef92f3b9cfd041f772c442f9R29
    ('C1CN[C@](O)(N)1',  'C1CN[C@]1(O)(N)'), # all toolkits agree
    ('C1CN[C@H]1O',      'C1CN[C@@H](O)1'),
    ('C1CN[C@]12(O).N2', 'C1CN[C@](O)12.N2'),
    ('C[C@]1(O)NCC1',    'C[C@@](O)1NCC1'),
    ('C[C@]1(NCC1)O',    'C[C@](NCC1)(O)1'),

    # ring case from OpenSMILES spec
    ('N1[C@H](Cl)[C@@H](Cl)C(Cl)CC1', 'ClC1CCN[C@H](Cl)[C@H]1Cl'),     # all toolkits agree
    ('Cl[C@@H]1[C@@H](Cl)C(Cl)CCN1', '[C@H](Cl)1[C@@H](Cl)C(Cl)CCN1'), # RDKit thinks these are different

    # dot disconnected stereo case all toolkits agree on
    ('Cl3.[C@]31(c2ccccc2)NCCCS1', 'Cl[C@]1(c2ccccc2)NCCCS1'),
]

for left, rght in SMILES_THAT_ARE_THE_SAME:
    rdleft = Chem.CanonSmiles(left)
    rdrght = Chem.CanonSmiles(rght)
    if rdleft != rdrght:
        print("RDKit failed to recognize these as the same:")
        print(left, '\t->\t', rdleft)
        print(rght, '\t->\t', rdrght)
```

Happy to turn the above into a proper RDKit unit test case once the fix is architected onto a branch somewhere.

### PR DESCRIPTION
**Pull request by coleb**
#### Reference Issue

Fixes #1652

#### What does this implement/fix? Explain your changes.

Fixes the issue with improperly parsing the stereochemistry of the first atom in the SMILES string when that atom is in a ring and its ring closure digit comes after a branch to a constituent. This is really a continuation of the fix done for issue #786 in pull request #823. 

#### Any other comments?

This should have fairly minimal effect unless some other tool somewhere created these weird looking SMILES. Even if some other tool decided that chiral atoms were a good place to start the SMILES (a reasonable decision), it would still likely put the ring closure digit immediately afterwards. 

```
>>> Chem.CanonSmiles('[C@@](F)1(C)CCO1')
'C[C@]1(F)CCO1'
>>> Chem.CanonSmiles('[C@]1(F)(C)CCO1')
'C[C@]1(F)CCO1'
```

### CODE REVIEW

File: Code/GraphMol/SmilesParse/SmilesParseOps.cpp [hunk 1]
```diff
@@ -25,8 +25,12 @@ void CheckRingClosureBranchStatus(RDKit::Atom *atom, RDKit::RWMol *mp) {
...
   // detect the branch (= atom isn't the last one)
   // and reverse the stereochem if the atom has chiral stereochemistry
-  // and is currently degree two (protects against C1CN[C@](O)(N)1)
-  if (atom->getIdx() != mp->getNumAtoms(true) - 1 && atom->getDegree() == 2 &&
+  // and is currently degree one   [C@@](F)1(C)CCO1
+  //     or currently degree two   C1CN[C@](O)(N)1
+  //        and not the first atom [C@](Cl)(F)1CC[C@H](F)CC1
+  if (atom->getIdx() != mp->getNumAtoms(true) - 1 &&
+      (atom->getDegree() == 1 ||
+       (atom->getDegree() == 2 && atom->getIdx() != 0)) &&
```
Comment (bp-kelley) [inline Code/GraphMol/SmilesParse/SmilesParseOps.cpp]: Isn't the first example is degree three (no hydrogens) no?

This one: [C@@](F)1(C)CCO1

Comment (coleb) [reply Code/GraphMol/SmilesParse/SmilesParseOps.cpp]: The way to interpret the comment is, 'the degree when the closure digit is encountered during parsing'.

```[C@@](F)1(C)CCO1```

So when the `1` digit is encountered in this instance there is only one neighbor to the carbon atom `[C@@](F)`.

CheckRingClosureBranchStatus is called directly by the yacc parser in blocks like this: 

```
| mol ring_number {
  RWMol * mp = (*molList)[$$];
  Atom *atom=mp->getActiveAtom();
  mp->setAtomBookmark(atom,$2);

  Bond *newB = mp->createPartialBond(atom->getIdx(),
                     Bond::UNSPECIFIED);
  mp->setBondBookmark(newB,$2);
  newB->setProp(RDKit::common_properties::_unspecifiedOrder,1);

  SmilesParseOps::CheckRingClosureBranchStatus(atom,mp);

  INT_VECT tmp;
  atom->getPropIfPresent(RDKit::common_properties::_RingClosures,tmp);
  tmp.push_back(-($2+1));
  atom->setProp(RDKit::common_properties::_RingClosures,tmp);

}
```

The other tokens that call it are: 

```
| mol BOND_TOKEN ring_number {
...
| mol MINUS_TOKEN ring_number {
```

Comment (bp-kelley) [reply Code/GraphMol/SmilesParse/SmilesParseOps.cpp]: @greglandrum Could we make the comment more explicit in the code?  I understand the intent better now, but @coleb had to explain it to me.

Comment (greglandrum) [reply Code/GraphMol/SmilesParse/SmilesParseOps.cpp]: Makes sense. What about:
```
  // github #786 and #1652: if the ring closure comes after a branch,
  // the stereochem is wrong.
  // This function is called while after closing a branch during construction of
  // the molecule from SMILES and corrects for what happens when parsing odd
  // (and arguably wrong) SMILES constructs like :
  //   1) [C@@](F)1(C)CCO1
  //   2) C1CN[C@](O)(N)1
  //   3) [C@](Cl)(F)1CC[C@H](F)CC1
  // In each of the first two cases the stereochemistry at the chiral atom
  // needs to be reversed. In the third case the stereochemistry should be
  // reversed when the Cl is added, but left alone with the F is added.
  // We recognize these situations using the degree and index of the chiral atom
```

Comment (coleb) [reply Code/GraphMol/SmilesParse/SmilesParseOps.cpp]: I've updated the comment to the above while fixing up the english a little and expounding a tad bit at the end. Let me know what you think.

**Comment by greglandrum** [comment]
That looks great.

File: Code/GraphMol/SmilesParse/SmilesParseOps.cpp (hunk 1 above)
Comment (bp-kelley) [reply Code/GraphMol/SmilesParse/SmilesParseOps.cpp]: @greglandrum @coleb looks good to me.

Comment (coleb) [reply Code/GraphMol/SmilesParse/SmilesParseOps.cpp]: Merge please, actually waiting on this one to push some other work forward. :-)

**Comment by greglandrum** [comment]
your wish is my command. ;-)

Thanks for the bug fix

**Pull request merged**